src/ssh/test-sshbridge.c

```diff
@@ -0,0 +1,1358 @@
+  GPid mock_sshd;
+  guint16 ssh_port;
+
+} TestCase;
+
+typedef struct {
+    const char *ssh_command;
+    const char *mock_sshd_arg;
+    const char *client_password;
+    const char *username;
+    const char *knownhosts_data;
+    const char *knownhosts_file;
+    const char *config;
+    const char *problem;
+    gboolean allow_unknown;
+} TestFixture;
+
+static GString *
+read_all_into_string (int fd)
+{
+  GString *input = g_string_new ("");
+  gsize len;
+  gssize ret;
+
+  for (;;)
+    {
+      len = input->len;
+      g_string_set_size (input, len + 256);
+      ret = read (fd, input->str + len, 256);
+      if (ret < 0)
+        {
+          if (errno != EAGAIN)
+            {
+              g_critical ("couldn't read from mock input: %s", g_strerror (errno));
+              g_string_free (input, TRUE);
+              return NULL;
+            }
+        }
+      else if (ret == 0)
+        {
+          return input;
+        }
+      else
+        {
+          input->len = len + ret;
+          input->str[input->len] = '\0';
+        }
+    }
+}
+
+static void
+spawn_setup (gpointer data)
+{
+  int fd = GPOINTER_TO_INT (data);
+
+  /* Send this signal to all direct child processes, when bridge dies */
+  prctl (PR_SET_PDEATHSIG, SIGHUP);
+
+  if (dup2 (fd, 0) < 0 || dup2 (fd, 1) < 0)
+    g_assert_not_reached ();
+
+  close (fd);
+}
+
+static void
+setup_mock_sshd (TestCase *tc,
+                 gconstpointer data)
+{
+  const TestFixture *fixture = data;
+  GError *error = NULL;
+  GString *port;
+  gchar *endptr;
+  guint64 value;
+  gint out_fd;
+
+  const gchar *argv[] = {
+      BUILDDIR "/mock-sshd",
+      "--user", g_get_user_name (),
+      "--password", PASSWORD,
+      fixture->mock_sshd_arg,
+      NULL
+  };
+
+  g_spawn_async_with_pipes (BUILDDIR, (gchar **)argv, NULL, G_SPAWN_DO_NOT_REAP_CHILD, NULL, NULL,
+                            &tc->mock_sshd, NULL, &out_fd, NULL, &error);
+  g_assert_no_error (error);
+
+  /*
+   * mock-sshd prints its port on stdout, and then closes stdout
+   * This also lets us know when it has initialized.
+   */
+
+  port = read_all_into_string (out_fd);
+  g_assert (port != NULL);
+  close (out_fd);
+  g_assert_no_error (error);
+
+  g_strstrip (port->str);
+  value = g_ascii_strtoull (port->str, &endptr, 10);
+  if (!endptr || *endptr != '\0' || value == 0 || value > G_MAXUSHORT)
+      g_critical ("invalid port printed by mock-sshd: %s", port->str);
+
+  tc->ssh_port = (gushort)value;
+  g_string_free (port, TRUE);
+}
+
+static const TestFixture fixture_mock_echo = {
+  .ssh_command = BUILDDIR "/mock-echo"
+};
+
+static const TestFixture fixture_cat = {
+  .ssh_command =  SRCDIR "/src/ws/mock-cat-with-init"
+};
+
+static gchar **
+setup_env (const TestFixture *fix)
+{
+  const gchar *command;
+  const gchar *knownhosts_file;
+  const gchar *config;
+  gchar **env = g_get_environ ();
+
+  config = fix ? fix->config : NULL;
+  if (!config)
+    config = SRCDIR "/src/ssh/mock-config";
```

Comment:
I sort of prefer it separate rather than having multiple ? .. : .. on a single line.